runtime/lua/vim/lsp/callbacks.lua

```diff
@@ -104,6 +92,23 @@ M['textDocument/publishDiagnostics'] = function(_, _, result)
     end
   end
 
+  util.buf_clear_diagnostics(bufnr)
+
+  -- Unloaded buffers should not handle diagnostics.
+  --    When the buffer is loaded, we'll call on_attach, which sends textDocument/didOpen.
+  --    This should trigger another publish of the diagnostics.
+  --
+  -- In particular, this stops a ton of spam when first starting a server for current
+  -- unloaded buffers.
+  if not api.nvim_buf_is_loaded(bufnr) then
+    -- Still save the diagnostics
+    -- Language servers may report compile or build errors via diagnostics
+    -- Users should be able to find these, even if they're in files which
+    -- are not loaded.
+    util.buf_diagnostics_save_positions(bufnr, result.diagnostics)
+    return
+  end
+
```

Comment:
We now save the diagnostics twice?

Reply:
It returns if the buffer is not loaded and if the buffer is loaded the `save_positions` call further above doesn't happen.

Reply:
Ah, yes, that's obviously correct. Could we instead just pull the shared `util.buf_diagnostics_save_positions(bufnr, result.diagnostics)` above this `if` block and remove it from line 112? I'd prefer to not have it copied between the two places, if that makes sense.